api/src/main/java/org/apache/iceberg/expressions/CountNonNull.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.expressions;
+
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.types.Types;
+
+public class CountNonNull<T> extends CountAggregate<T> {
+  private final int fieldId;
+
+  protected CountNonNull(BoundTerm<T> term) {
+    super(Operation.COUNT, term);
+    Types.NestedField field = term.ref().field();
+    this.fieldId = field.fieldId();
+  }
+
+  @Override
+  protected Long countFor(StructLike row) {
+    return term().eval(row) != null ? 1L : 0L;
+  }
+
+  @Override
+  protected Long countFor(DataFile file) {
+    // NaN value counts were not required in v1 and were included in value counts
+    return safeAdd(safeGet(file.valueCounts(), fieldId), safeGet(file.nanValueCounts(), fieldId, 0L));
```

Comment:
Shall we subtract the `nullValueCounts`?

Reply:
Yes, you're right. This will include NaN and null values:
> Map from column id to number of values in the column (including null and NaN values)

That means we should actually not add the NaN count.